expression/builtin_other.go

```diff
@@ -688,13 +689,26 @@ func (b *builtinSetVarSig) evalString(row chunk.Row) (res string, isNull bool, e
 	if isNull || err != nil {
 		return "", isNull, err
 	}
-	res, isNull, err = b.args[1].EvalString(b.ctx, row)
+
+	datum, err := b.args[1].Eval(row)
+	isNull = datum.IsNull()
 	if isNull || err != nil {
 		return "", isNull, err
 	}
+
+	res, err = datum.ToString()
+	if err != nil {
+		return "", isNull, err
+	}
+
 	varName = strings.ToLower(varName)
 	sessionVars.UsersLock.Lock()
-	sessionVars.Users[varName] = stringutil.Copy(res)
+	if len(datum.Collation()) > 0 {
```

Comment:
The code here is quite similar to https://github.com/pingcap/tidb/pull/15449/files#diff-3b5e42b4b881a9e47146b5ae9b58fad2R97, can you wrap a function for them?

Reply:
done